src/MSBuild.Sdk.SqlProj.Aspire/PublishSqlProjectLifecycleHook.cs

```diff
@@ -1,67 +1,173 @@
-                    state => state with { State = new ResourceStateSnapshot("Failed", KnownResourceStateStyles.Error) });
-                continue;
-            }
+        return Task.CompletedTask;
+    }
+
+    private void StartDacpacDeployment(SqlProjectResource sqlProject, DistributedApplicationModel application, CancellationToken cancellationToken)
+    {
+        var logger = _resourceLoggerService.GetLogger(sqlProject);
 
-            var targetDatabaseResourceName = sqlProject.Annotations.OfType<TargetDatabaseResourceAnnotation>().Single().TargetDatabaseResourceName;
-            var targetDatabaseResource = application.Resources.OfType<SqlServerDatabaseResource>().Single(r => r.Name == targetDatabaseResourceName);
-            var connectionString = await targetDatabaseResource.ConnectionStringExpression.GetValueAsync(cancellationToken);
-            if (connectionString == null)
+        _ = Task.Run(async () =>
+        {
+            int retryCount = 10;
+            int delayInMillisecodns = 1000;
+
+            for (int i = 0; i < retryCount; i++)
             {
-                logger.LogError("Failed to retrieve connection string for target database {TargetDatabaseResourceName}.", targetDatabaseResourceName);
-                await _resourceNotificationService.PublishUpdateAsync(sqlProject,
-                    state => state with { State = new ResourceStateSnapshot("Failed", KnownResourceStateStyles.Error) });
-                continue;
+                try
+                {
+                    var dacpacPath = sqlProject.GetDacpacPath();
+                    if (!File.Exists(dacpacPath))
+                    {
+                        logger.LogError("SQL Server Database project package not found at path {DacpacPath}.", dacpacPath);
+
+                        await _resourceNotificationService.PublishUpdateAsync(sqlProject,
+                            state => state with { State = new ResourceStateSnapshot(KnownResourceStates.FailedToStart, KnownResourceStateStyles.Error) });
+                        break;
+                    }
+
+                    var targetDatabaseResourceName = sqlProject.Annotations.OfType<TargetDatabaseResourceAnnotation>().Single().TargetDatabaseResourceName;
+                    var targetDatabaseResource = application.Resources.OfType<SqlServerDatabaseResource>().Single(r => r.Name == targetDatabaseResourceName);
+                    var targetServerResource = targetDatabaseResource.Parent;
+                    var connectionString = await targetDatabaseResource.ConnectionStringExpression.GetValueAsync(cancellationToken);
+
+                    var serverReadyAnnotation = await CheckServerReady(sqlProject, logger, targetDatabaseResourceName, targetServerResource, connectionString);
+
+                    await Publish(sqlProject, logger, dacpacPath, targetDatabaseResource, connectionString, serverReadyAnnotation, cancellationToken);
+                }
+                catch
```

Comment:
Currently we wait for timeout of the retries on any exceptions - until we know more that seems like an OK approach to me

Reply:
@jmezach Thoughts?

Reply:
Yeah, I think that makes sense. We might want to write the exception to the log though when we give up, so that we have some diagnostics to further fine tune this in the future.